src/Features/LanguageServer/CommonLanguageServerProtocol.Framework/IRequestHandler.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+#nullable enable
+
+namespace CommonLanguageServerProtocol.Framework;
+
+/// <summary>
+/// Top level type for LSP request handler.
+/// </summary>
+public interface IRequestHandler
+{
+    /// <summary>
+    /// Whether or not the solution state on the server is modified
+    /// as a part of handling this request.
+    /// </summary>
+    bool MutatesSolutionState { get; }
+
+    /// <summary>
+    /// Whether or not the handler execution queue should build a solution that represents the LSP
+    /// state of the world. If this property is not set <see cref="RequestContext.Solution"/> will be <see langword="null"/>
+    /// and <see cref="RequestContext.Document"/> will be <see langword="null"/>, even if <see cref="IRequestHandler{RequestType, ResponseType, RequestContextType}.GetTextDocumentUri(RequestType)"/>
+    /// doesn't return null. Handlers should still provide text document information if possible to
+    /// ensure the correct workspace is found and validated.
+    /// </summary>
+    bool RequiresLSPSolution { get; }
+}
+
+public interface IRequestHandler<RequestType, ResponseType, RequestContextType> : IRequestHandler
+{
+    /// <summary>
+    /// Gets the identifier of the document from the request, if the request provides one. Can be Uri, TextDocumentIdentifier, or anything else your RequestContextFactory can handle.
+    /// </summary>
+    object? GetTextDocumentUri(RequestType request);
```

Comment:
feels like this shoudl be a type parameter too to be strongly typed.  Should we just manduate thsi is a URI?  what's the benefit of a TextDocumentIdentifier over that for example?

Reply:
We cannot mandate that this be a URI unfortunately due to [the issue discussed here](https://github.com/dotnet/roslyn/pull/63149#discussion_r936117489) (I should change the name to reflect that).

I like the idea to make this a type parameter, but I'm cautious since there are already so many. If people follow the pattern I've [established here](https://github.com/dotnet/roslyn/pull/62924/files#diff-1b890f5c853317d596952acabc3683b2c83893e0b352b58a1c26d7443fecbe13) for their implementation they can avoid having to worry about it too much, but I figure it's worth discussing with people. Anyone have strong objections?